keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -0,0 +1,272 @@
+  examples include requests from an aggregated apiserver to create
+  TokenReview and SubjectAccessReview objects.  Today it is possible
+  that the very load imposed by request A crowds out requests B
+  involved in serving A.
+
+- Garbage Collector crowded out.  The garbage collector should keep up
+  with the workload, but in an overload situation today this is not
+  assured to happen.
+
+- Deployment of Doom.  We had a situation where a bug in the
+  Deployment controller caused it to run amuck under certain
+  circumstances, issuing requests in a tight loop.  We would like
+  controller bugs to not take the whole system down.
+
+- Kubelet Amuck.  The controller that runs amuck might not be a
+  central singleton, it could be a kubelet, kube-proxy, or other
+  per-node or otherwise multiplied controller.  In such a situtation
+  we would like only the guilty individual to suffer, not all its
+  peers and the rest of the system.
+
+- Overbearing or buggy tenants.  In a multi-tenant scenario, we would
+  like to prevent some tenants from crowding out the others.  Various
+  usage scenarios involve identifying the tenant in the following
+  ways.
+  
+  - Each tenant corresponds with a kube API namespace.
+
+  - Each tenant corresponds with a user name.
+
+  - Each tenant corresponds with a prefix of the user name.
+
+  - Each tenant corresponds with a user's group.  Other groups may
+    exist.  There is a subset of the groups that serve to identify
+    tenants.  Each user belongs to exactly one of the
+    tenant-identifying groups.
+
+This KEP introduces new functionality in apiservers, and it should be
+possible to monitor this functionality through Prometheus metrics
+available from the apiservers.
+
+This KEP introduces new configuration objects, and they really
+matter; it should be easy to apply suitable access controls.
+
+There should be some reasonable defaults.
+
+### Non-Goals
+
+This will be our first cut at a significant area of functionality, and
+our goals are deliberately modest.  We want to implement something
+useful but limited and get some experience before going further.  Our
+precise modesty has not been fully agreed.  Following is an initial
+stake in the ground.
+
+- No coordination between apiservers nor with a load balancer is
+  attempted.  In this KEP each apiserver independently protects
+  itself.  We imagine that later developments may add support for
+  informing load balancers about the load state of the apiservers.
+
+- The fairness does not have to be highly precise.  Any rough fairness
+  will be good enough.
+
+- WATCH and CONNECT requests are out of scope.  These are of a fairly
+  different nature than the others, and their management will be more
+  complex.  Also they are arguably less of an observed problem.
+
+- We are only concerned with protection of the CPU and memory of the
+  apiserver.  We are not concerned with etcd performance, nor output
```

Comment:
Perhaps:

We are only concerned with protection of the CPU and memory of the apiserver consumed by servicing requests. We deliberately exclude memory involved in queueing requests and reading their headers. The intention is to perform adequately under a very high but reasonable request load, not to survive a DDoS attack.

Reply:
ah I see most of that is covered below, might still be worth it to mention the memory consumed by queueing / reading headers.

Reply:
I think we have not agreed to ignore memory consumed by queued requests.  In fact, some limitation on quantity queued is in the latest proposal.

I think the exclusion of memory in unreachable lower levels of the stack, and memory released before this handler runs, is only a matter of what can practically be implemented; it is not a matter of lack of desire to protect it.